### PR TITLE
Feat/schemas

### PR DESCRIPTION
**Pull request by khvn26**
Contributes to https://github.com/Flagsmith/flagsmith/issues/2079

### CODE REVIEW

**Review by gagantrivedi — CHANGES_REQUESTED** [review]

File: .gitmodules
Comment (khvn26) [reply .gitmodules:3]: ssh-based submodule didn't play well with `pip` in GH pipelines in dowstream PRs (https://github.com/Flagsmith/flagsmith/pull/2290, https://github.com/Flagsmith/edge-api/pull/153 and https://github.com/Flagsmith/flagsmith-engine/pull/167).
This will be removed before merge.

File: flag_engine/features/models.py
Comment (gagantrivedi) [inline flag_engine/features/models.py:32]: Should we use `conint` here to make sure it is a positive integer?

File: flag_engine/features/models.py
Comment (gagantrivedi) [inline flag_engine/features/models.py:63]: `.. or equal to..`

File: flag_engine/features/models.py
Comment (gagantrivedi) [inline flag_engine/features/models.py:71]: I think this is a bit odd that validation happens after the object is modified, e.g. if append raises an error on `some object`  I would not expect that object to be modified  - Does that make sense?

**Review by gagantrivedi — DISMISSED** [review]

File: flag_engine/features/models.py
Comment (khvn26) [reply flag_engine/features/models.py:74]: Agreed, fixed for both custom collections.

File: flag_engine/features/models.py
Comment (khvn26) [reply flag_engine/features/models.py:32]: Why, though? Current schemas don't support that constraint, do they?

File: flag_engine/features/models.py
Comment (khvn26) [reply flag_engine/features/models.py]: Fixed.

File: flag_engine/features/models.py
Comment (matthewelwell) [inline flag_engine/features/models.py:48]: Just to confirm here, as far as I'm aware the `init=False` argument to the `field()` function means that the value is not a valid argument to `__init__()`, allowing us to force users to set the value using the `set_value()` method. I'm not _entirely_ sure why we did this but I want to make sure that we're not losing/breaking any functionality here?

File: flag_engine/features/models.py
Comment (matthewelwell) [inline flag_engine/features/models.py:94]: Just checking whether this should (strictly) be: 

```suggestion
    def get_value(self, identity_id: typing.Union[type(None), int, str] = None) -> typing.Any:
```

File: flag_engine/identities/builders.py
Comment (matthewelwell) [inline flag_engine/identities/builders.py:7]: This is so much nicer :ok_hand:

File: flag_engine/identities/models.py
Comment (matthewelwell) [inline flag_engine/identities/models.py:16]: Just a general comment here that I definitely prefer the `marshmallow` `@validates_schema` decorator for providing custom decorators! Maybe this is more explicit though.

File: flag_engine/identities/models.py
Comment (matthewelwell) [inline flag_engine/identities/models.py:28]: Perhaps python will handle some optimisations for this internally but it certainly feels like we could optimise this by generating the list of feature ids up front?

File: flag_engine/identities/models.py
Comment (matthewelwell) [inline flag_engine/identities/models.py:52]: What's the benefit to using a concrete field vs a property? 

It feels kinda wrong that we're setting an attribute on the model in a validator, is this a valid pattern in pydantic or are we _slightly_ abusing the `root_validator` decorator here? Wouldn't this better live in an overridden `__init__` method (if we want to keep it as a concrete attr vs a property).

File: flag_engine/identities/traits/models.py
Comment (matthewelwell) [inline flag_engine/identities/traits/models.py:8]: What does the `...` signify here? Why not use `None`?

File: flag_engine/segments/types.py
Comment (matthewelwell) [inline flag_engine/segments/types.py:1]: Oh nice, I didn't know you could do this!

File: tests/end_to_end/test_flag_engine_end_to_end.py
Comment (matthewelwell) [inline tests/end_to_end/test_flag_engine_end_to_end.py:35]: It makes me a little nervous throwing away these tests. Are they covered in the Core API PR perhaps?

File: tests/unit/features/test_features_models.py
Comment (matthewelwell) [inline tests/unit/features/test_features_models.py:55]: Should there be a `"text"` or `"body"` key here as well that we can check?

File: tests/unit/identities/test_identities_models.py
Comment (matthewelwell) [inline tests/unit/identities/test_identities_models.py:120]:
```suggestion
def test_identity_model__identity_features__append__expected_result(
```

File: tests/unit/identities/traits/test_models.py
Comment (matthewelwell) [inline tests/unit/identities/traits/test_models.py:9]: 😂

**Review by matthewelwell — COMMENTED** [review]
image

Love this ^ 👌 

I've added a few comments but, on the whole, it's really great to see where we're seeing the benefit from this refactor. It's harder to see in the other 2 PRs but you can definitely see it here!

File: flag_engine/environments/models.py
Comment (khvn26) [reply flag_engine/environments/models.py:52]: Done.

File: tests/end_to_end/test_flag_engine_end_to_end.py
Comment (khvn26) [reply tests/end_to_end/test_flag_engine_end_to_end.py:35]: Indeed these cases are covered in the Core API PR, and we should have more confidence in them because they use real Django model instances.

File: flag_engine/features/models.py
Comment (khvn26) [reply flag_engine/features/models.py:32]: I can think of a couple — minimize the diff, try to reach 1:1 feature parity with the current engine version.
We have a typing PR #168 where we can address this.

File: flag_engine/environments/integrations/models.py
Comment (khvn26) [reply flag_engine/environments/integrations/models.py:7]: I prefer to reproduce the initial schemas' functionality here.

File: flag_engine/identities/traits/models.py
Comment (khvn26) [reply flag_engine/identities/traits/models.py:8]: `...` signifies a required `__init__` argument. Even when we're setting the value to `None`, we're required to do this explicitly. This was implemented in the Marshmallow schema too.

File: flag_engine/identities/models.py
Comment (khvn26) [reply flag_engine/identities/models.py:16]: After switching to Pydantic v2 we'll be able to use a custom `List` subclass instead of this hacky collection model.

Or we could (something I'd prefer to do) not use custom collections altogether and move their validation/interfaces to the parent model.

File: flag_engine/identities/models.py
Comment (khvn26) [reply flag_engine/identities/models.py:28]: The point of those partial iterations are to avoid iterating over the full list in case duplicates are found. I lifted this approach from Pydantic's `ConstrainedList.unique_items_validator` implementation. Generating the feature id list would negate the optimisation. Maybe it's redundant for our data and we could simplify this code to iterate over all the ids right away and only do the membership check once.

File: flag_engine/segments/types.py
Comment (khvn26) [reply flag_engine/segments/types.py:1]: Pydantic is awesome :)

File: tests/unit/features/test_features_models.py
Comment (khvn26) [reply tests/unit/features/test_features_models.py:55]: Do you mean `msg`? The message text is tested in `test_feature_state_model__multivariate_feature_state_values__append_invalid__raises_expected`.

File: flag_engine/features/models.py
Comment (matthewelwell) [reply flag_engine/features/models.py:32]: Ok.

File: flag_engine/environments/models.py
Comment (matthewelwell) [reply flag_engine/environments/models.py:52]: Maybe it hasn't been pushed yet but it doesn't look done?

File: flag_engine/identities/models.py
Comment (matthewelwell) [reply flag_engine/identities/models.py:28]: Ok, yeah, that makes sense. I guess, realistically this list is going to be a few items long in 99% of cases so we just shouldn't care about it either way :D

File: flag_engine/environments/models.py
Comment (khvn26) [reply flag_engine/environments/models.py:52]: It is now. Sorry for the misfire

File: flag_engine/features/models.py
Comment (khvn26) [reply flag_engine/features/models.py]: Fixed.

File: tests/unit/features/test_features_models.py
Comment (khvn26) [reply tests/unit/features/test_features_models.py:55]: Agreed, done.

File: flag_engine/features/models.py
Comment (matthewelwell) [reply flag_engine/features/models.py:48]: @khvn26 I think this question is still outstanding here?

**Review by matthewelwell — COMMENTED** [review]
No additional comments here - I've resolved most of the conversations that were outstanding but I think there are still a few left waiting for a response.

File: flag_engine/features/models.py
Comment (khvn26) [reply flag_engine/features/models.py:97]: It should not https://docs.python.org/3.7/library/typing.html#type-aliases

File: flag_engine/identities/models.py
Comment (khvn26) [reply flag_engine/identities/models.py:20]: Done.

File: flag_engine/features/models.py
Comment (khvn26) [reply flag_engine/features/models.py:48]: We're not breaking anything, only adding an ability to set `feature_state_value` via `__init__` as a usual dataclass attribute.

The `set_value` method is kept as it was for backwards compatibility.

Comment (khvn26) [reply flag_engine/features/models.py:48]: The `set_value` method is a legacy interface initially conceived to get the feature state value that was private.
To simplify serialization it was made public, and now this method is redundant.
It's still being kept for backwards compatibility but we'll steer clear of using it in the future and deprecate it eventually.

File: .gitmodules
Comment (khvn26) [reply .gitmodules:3]: @matthewelwell encountered the same problem and we decided to leave the HTTPS submodule.

**Comment by khvn26** [comment]
Merge this after https://github.com/Flagsmith/flagsmith-engine/pull/171 is merged and released.

**Review by matthewelwell — APPROVED** [review]

**Pull request merged**
